### PR TITLE
sql: support array_in and regproc equality

### LINKED ISSUE #7790: pgcompat: support `regproc` object identifier type

### PR DESCRIPTION
**Pull request by mjibson**
This enables pgjdbc's getSQLType method.

Fixes #7790

### CODE REVIEW

File: test/lang/java/smoketest/SmokeTest.java [hunk 1]
```diff
@@ -73,4 +75,12 @@ void testBinaryTimestamp() throws SQLException, ClassNotFoundException {
...
         stmt.close();
     }
+
+    @Test
+    void testGetSqlType() throws SQLException, ClassNotFoundException {
+        Class.forName("org.postgresql.core.BaseConnection");
+        TypeInfoCache ic = new TypeInfoCache(conn.unwrap(org.postgresql.core.BaseConnection.class), 0);
+        Assertions.assertEquals(ic.getSQLType("int2"), Types.SMALLINT);
+        Assertions.assertEquals(ic.getSQLType("_int2"), Types.ARRAY);
+    }
```
Comment (benesch) [inline test/lang/java/smoketest/SmokeTest.java:85]: So good!

File: src/coord/src/catalog/builtin.rs [hunk 2]
```diff
@@ -1261,14 +1263,19 @@ pub const PG_TYPE: BuiltinView = BuiltinView {
         0
     )
         AS typarray,
+    -- Unfortunately we cannot use 'array_in'::pg_catalog.regproc, so hard code the
+    -- array_in oid.
```
Comment (benesch) [inline src/coord/src/catalog/builtin.rs:1267]: Wait, why not?

File: src/coord/src/catalog/builtin.rs [hunk 3]
```diff
@@ -1245,7 +1245,9 @@ pub const PG_TYPE: BuiltinView = BuiltinView {
     mz_types.oid,
     mz_types.name AS typname,
     mz_schemas.oid AS typnamespace,
-    typtype,
+    -- 'a' is used internall to denote an array type, but in postgres they show up
```
Comment (sploiselle) [inline src/coord/src/catalog/builtin.rs:1248]: s/internall/internally/ but not worth going back through CI for

**Review by benesch — APPROVED** [review]

**Review by sploiselle — APPROVED** [review]

File: src/coord/src/catalog/builtin.rs (hunk 2 above)
Comment (mjibson) [reply src/coord/src/catalog/builtin.rs:1267]: Unsure. My best guess is initialization order? The text -> regproc conversion uses pg_proc which maybe doesn't exist yet? I could attempt to track it down.

Comment (mjibson) [reply src/coord/src/catalog/builtin.rs:1267]: For example, using the text version here works after mz has initialized.

**Pull request merged**